### PR TITLE
BUG: raise when sort_index with ascending=None

### LINKED ISSUE #39434: BUG: Regression in sort_index with ascending=None
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
# Your code here
import pandas
import numpy as np
df = pandas.DataFrame(np.arange(64))
length = len(df.index)
df.index = [(i - length / 2) % length for i in range(length)]
df.sort_index(axis=0, ascending=None, na_position="first")
ValueError: the 'axis' parameter is not supported in the pandas implementation of argsort()
```

#### Problem description

There was no such issue in pandas version 1.1.5

#### Output of ``pd.show_versions()``

pandas           : 1.2.1
numpy            : 1.19.5
pytz             : 2020.5
dateutil         : 2.8.1
pip              : 20.3.3
setuptools       : 51.1.2.post20210112
Cython           : None
pytest           : 6.2.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : 0.4.1
xlsxwriter       : None
lxml.etree       : 4.6.2
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.12.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : None
fsspec           : 0.8.3
fastparquet      : None
gcsfs            : None
matplotlib       : 3.2.2
numexpr          : 2.7.2
odfpy            : None
openpyxl         : 3.0.6
pandas_gbq       : None
pyarrow          : 2.0.0
pyxlsb           : None
s3fs             : 0.5.2
scipy            : 1.6.0
sqlalchemy       : 1.3.22
tables           : 3.6.1
tabulate         : None
xarray           : 0.16.2
xlrd             : 2.0.1
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by ivanovmg**
- [ ] closes #39434
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

Added validation of ``ascending`` for ``sort_index`` methods.
Proposed a solution to raise ``ValueError`` when either of the elements of ``ascending`` turned out to be ``None``.

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -4533,6 +4533,8 @@ def sort_index(
 
         inplace = validate_bool_kwarg(inplace, "inplace")
         axis = self._get_axis_number(axis)
+        ascending = validate_ascending(ascending)
```
Comment (jreback) [inline pandas/core/generic.py:4536]: you don't need allow of this, just call `validate_bool_kwarg`

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 1 above)
Comment (ivanovmg) [reply pandas/core/generic.py:4536]: ``validate_bool_kwargs`` accepts ``None`` as a valid value.
Do we consider ``ascending=None`` as valid?
https://github.com/pandas-dev/pandas/issues/39434#issuecomment-768407714

Comment (jreback) [reply pandas/core/generic.py:4536]: in this case no
so i can enhance the checker with an arg to only allow an actual bool

Comment (jreback) [reply pandas/core/generic.py:4536]: pls use `validate_bool_kwarg` instead of adding *anything* here. you can add a kwargs to that function itself to make things strict

Comment (ivanovmg) [reply pandas/core/generic.py:4536]: ``ascending`` here may be a single value or a sequence of items.
I am afraid that I still need to add validation of either a scalar value or each of the items in the sequence.

Comment (simonjayhawkins) [reply pandas/core/generic.py:4536]: in the def we have `ascending: bool_t = True`, so should probably update that too as part of this PR.

File: pandas/util/_validators.py [hunk 2]
```diff
@@ -205,9 +205,39 @@ def validate_args_and_kwargs(fname, args, kwargs, max_fname_arg_count, compat_ar
     validate_kwargs(fname, kwargs, compat_args)
 
 
-def validate_bool_kwarg(value, arg_name):
-    """ Ensures that argument passed in arg_name is of type bool. """
-    if not (is_bool(value) or value is None):
+def validate_bool_kwarg(value, arg_name, none_allowed=True, int_allowed=False):
```
Comment (simonjayhawkins) [inline pandas/util/_validators.py:208]: if `validate_bool_kwarg` allowed ints through before, wouldn't `int_allowed` need to default to True for backwards compatibility?

File: pandas/util/_validators.py [hunk 3]
```diff
@@ -205,9 +205,39 @@ def validate_args_and_kwargs(fname, args, kwargs, max_fname_arg_count, compat_ar
...
+    """
+    Ensure that argument passed in arg_name can be interpreted as boolean.
+
+    Parameters
+    ----------
+    value : bool
+        Value to be validated.
+    arg_name : str
+        Name of the argument. To be reflected in the error message.
+    none_allowed : bool, optional
```
Comment (simonjayhawkins) [inline pandas/util/_validators.py:218]: add default

File: pandas/util/_validators.py [hunk 4]
```diff
@@ -205,9 +205,39 @@ def validate_args_and_kwargs(fname, args, kwargs, max_fname_arg_count, compat_ar
...
+        Whether to consider None to be a valid boolean.
+    int_allowed : bool, optional
```
Comment (simonjayhawkins) [inline pandas/util/_validators.py:220]: same

File: pandas/core/generic.py [hunk 5]
```diff
@@ -11772,3 +11774,12 @@ def _doc_params(cls):
     The required number of valid values to perform the operation. If fewer than
     ``min_count`` non-NA values are present the result will be NA.
 """
+
+
+def validate_ascending(ascending):
```
Comment (jreback) [inline pandas/core/generic.py:11779]: can you move to pandas/util/validators.py

shouldn't this raise for a list-like case if something is wrong?

Comment (jreback) [reply pandas/core/generic.py:11779]: e.g. this shouldn't return anything

pls add some typing

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/util/_validators.py (hunk 2 above)
Comment (ivanovmg) [reply pandas/util/_validators.py:208]: I checked ``is_bool``, and by default it considers ``int`` as a non-boolean.

```
In [1]: from pandas.core.dtypes.common import is_bool

In [2]: is_bool(1)
Out[2]: False
```

File: pandas/core/generic.py (hunk 5 above)
Comment (ivanovmg) [reply pandas/core/generic.py:11779]: It raises for a list-like case if any of the items in ``ascending`` list is not treated as a valid boolean.
This construct does just that.
```
return [validate_bool_kwarg(item, "ascending", **kwargs) for item in ascending]
```

Comment (jreback) [reply pandas/core/generic.py:11779]: this can simply be part of validate_bool_kwarg or another function but should be co-located with validate_bool_kwarg

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you merge master and fixup to comments.

File: pandas/core/generic.py (hunk 1 above)
Comment (ivanovmg) [reply pandas/core/generic.py:4540]: Extended typing for ``ascending``

File: pandas/util/_validators.py (hunk 3 above)
Comment (ivanovmg) [reply pandas/util/_validators.py]: Added

File: pandas/util/_validators.py (hunk 4 above)
Comment (ivanovmg) [reply pandas/util/_validators.py]: Added

File: pandas/core/generic.py (hunk 5 above)
Comment (ivanovmg) [reply pandas/core/generic.py]: Moved the function to ``pandas/util/_validators.py``.
It is not clear for me at the moment how to make this validation anyhow different (as a part of ``validate_bool_kwarg``) due to the fact that multiple types are acceptable for ``ascending``.

File: pandas/core/series.py [hunk 6]
```diff
@@ -3039,7 +3041,7 @@ def update(self, other) -> None:
     def sort_values(
         self,
         axis=0,
-        ascending=True,
+        ascending: Union[Union[bool, int], Sequence[Union[bool, int]]] = True,
```
Comment (jreback) [inline pandas/core/series.py:3044]: Union[bool, Sequence[bool]]. int is no-where in this. same for all the others.

File: pandas/core/series.py [hunk 7]
```diff
@@ -3231,7 +3234,7 @@ def sort_values(
 
         # GH 35922. Make sorting stable by leveraging nargsort
         values_to_sort = ensure_key_mapped(self, key)._values if key else self._values
-        sorted_index = nargsort(values_to_sort, kind, ascending, na_position)
+        sorted_index = nargsort(values_to_sort, kind, bool(ascending), na_position)
```
Comment (jreback) [inline pandas/core/series.py:3237]: why is this needed?

File: pandas/util/_validators.py (hunk 2 above)
Comment (jreback) [inline pandas/util/_validators.py:208]: remove the int_allowed (its not)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/methods/test_sort_index.py [hunk 8]
```diff
@@ -761,6 +761,22 @@ def test_sort_index_with_categories(self, categories):
...
 
+    @pytest.mark.parametrize(
+        "ascending",
+        [
+            None,
+            (True, None),
+            (False, "True"),
+        ],
+    )
+    def test_sort_index_ascending_bad_value_raises(self, ascending):
```
Comment (simonjayhawkins) [inline pandas/tests/frame/methods/test_sort_index.py:772]: can you add the issue number as a comment

File: pandas/tests/frame/methods/test_sort_index.py [hunk 9]
```diff
@@ -761,6 +761,22 @@ def test_sort_index_with_categories(self, categories):
         )
         tm.assert_frame_equal(result, expected)
 
+    @pytest.mark.parametrize(
+        "ascending",
+        [
+            None,
+            (True, None),
+            (False, "True"),
```
Comment (simonjayhawkins) [inline pandas/tests/frame/methods/test_sort_index.py:769]: can you make these lists

Comment (simonjayhawkins) [reply pandas/tests/frame/methods/test_sort_index.py:769]: and maybe add an empty list too

File: pandas/core/frame.py [hunk 10]
```diff
@@ -5560,7 +5560,7 @@ def sort_index(
         self,
         axis: Axis = 0,
         level: Optional[Level] = None,
-        ascending: bool = True,
+        ascending: Union[Union[bool, int], Sequence[Union[bool, int]]] = True,
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:5563]: also could change docstring `bool or list of bools` -> `bool or list-like of bools`

File: pandas/core/series.py (hunk 6 above)
Comment (simonjayhawkins) [reply pandas/core/series.py:3044]: and docstring L3062 `ascending : bool, default True` should be updated to match

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
@ivanovmg can you resolve conflicts and address comments

File: pandas/tests/frame/methods/test_sort_index.py (hunk 9 above)
Comment (ivanovmg) [reply pandas/tests/frame/methods/test_sort_index.py]: Currently an empty list does not raise an error, so the test fails.
Should I change the code, so that an empty list raises?

File: pandas/tests/frame/methods/test_sort_index.py (hunk 8 above)
Comment (ivanovmg) [reply pandas/tests/frame/methods/test_sort_index.py:772]: Done.

File: pandas/core/frame.py (hunk 10 above)
Comment (ivanovmg) [reply pandas/core/frame.py:5628]: Done.

File: pandas/core/series.py (hunk 7 above)
Comment (ivanovmg) [reply pandas/core/series.py:3297]: I did that in order to match the expected type for ``ascending`` in ``nargsort`` function.
It must be ``bool`` as per the function signature.
However, at this particular point ``ascending`` can be an integer as well (previously it was allowed).

File: pandas/util/_validators.py (hunk 2 above)
Comment (ivanovmg) [reply pandas/util/_validators.py:208]: This is required for ``validate_ascending`` as int is actually considered as a valid boolean there.
Previously I had this check in ``validate_ascending`` itself, but I was requested to use ``validate_boolean``, so I extended it in such a way.

It may be necessary to maybe restrict ``ascending`` to use of booleans only (no integers), but this is a breaking change and should probably be discussed first.

**Comment by jreback** [comment]
ok looks fine. can you add a whtasnew note for 1.2.3 in regression section. merge master and ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @ivanovmg

**Comment by jreback** [comment]
@meeseeksdev backport 1.2.x
